specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+    CircSupply abi.TokenAmount 
+    // Status of the checkpoint (`Active`, `Inactive`, etc.)
+    Status     Status
+    // Latest checkpoint committed for the subnet.
+    // Kept for verification purposes.
+    PrevCheckpoint schema.Checkpoint
+}
+```
+
+## Consensus Interface
+Each subnet can run its own implementation of a consensus algorithm. In order for different consensus implementations to operate semlessly with the Filecoin stack, we decouple the core methods of the current consensus in its own interace. In order for a consensus implementation to be usuable in a subnet it needs to implement this interface.
+> Consensus Interface
+```go
+type Consensus interface {
+    // Performs a complete check of a proposed block to either
+    // accept it or reject it.
+    ValidateBlock(ctx context.Context, b *types.FullBlock) (err error)
+    // Light check performed when a block is received through 
+    // the pubsub channel.
+    ValidateBlockPubsub(ctx context.Context, self bool, msg *pubsub.Message) (pubsub.ValidationResult, string)
+
+    // Used by mining processes to assemble and propose a new
+    // signed block.
+    CreateBlock(ctx context.Context, w api.Wallet, bt *api.BlockTemplate) (*types.FullBlock, error)
+    // Return the type of consensus. 
+    Type() hierarchical.ConsensusType
+}
+```
+An example of a BFT-like consensus protocol being integrated in HC through this consensus interface can be found [here](https://github.com/filecoin-project/mir)
+
+> TODO: Add a section describing how to integrate new consensus algorithms into HC.
+
+## Lifecycle of a subnet
+The following section presents an overview of the lifecycle of a subnet.
+
+### Spawning and joining a subnet
+Creating a new subnet instantiates a new independent state with all its subnet-specific requirements to operate independently. This includes, in particular: a new pubsub topic that peers use as the transport layer to exchange subnet-specific messages, a new mempool instance, a new instance of the Virtual Machine (VM), as well as any other additional module required by the consensus that the subnet is running (builtin-actors, mining resources, etc.). 
+
+To spawn a new subnet, peers need to deploy a new `SubnetActor` that implements the core logic for the governance of the new subnet. The contract specifies the consensus protocol to be run by the subnet and the set of policies to be enforced for new members, leaving members, checkpointing, killing the subnet, etc. For a new subnet to interact with the rest of the hierarchy, it needs to be registered in the `SCA` of the parent chain. The `SCA` is a system actor that exposes the interface for subnets to interact with the hierarchical consensus protocol. This smart contract includes all the available functionalities related to subnets and their management. And, as `SA`s are user-defined and untrusted, it also enforces security requirements, fund management and the cryptoeconomics of hierarchical consensus.
```

Comment:
Agree! Do you think that the description in `SubnetActor` and `SCA` are enough to understand the lifecycle of a subnet? Initially, I didn't include this section, but I was afraid that with all the methods and data structures, the high-level description would be lost. I appreciate any suggestion on how to improve this :)

Reply:
I found that on first read I found it easy enough to follow, but when I take a closer look I'm struggling a bit because:
* Data structures are mentioned in the interfaces, but only introduced much later when the text explains what they are about
* Interfaces have comments for methods about things like who can call them, but not what they do in detail - that comes later in prose, where the method is also just part of the text, so I have to search
* The raw Cid in the data structures doesn't tell much without comment, I have to find the field mentioned in the text
* I get confused by the parent vs child context when we just say "SCA"

This is obviously a huge spec and a lot of effort went into it, but it feels a little bit like a "report" of what has been done, rather than a spec of how someone should implement it from scratch. Examples of this are things that are mentioned in parentheses as "(after that other thing happens)".

Perhaps another option would be to:
1. Introduce the actor state first with all the supporting data structures, use as much typing annotation as possible
2. Introduce the methods/messages the actors can handle, systematically explaining things like
  * who can call them
  * what other preconditions apply
  * which parts of the state they touch and how do they change it
  * what are their invariants, if any
  * what messages they send to other actors

That way I should be able to look up an operation and compare it to the code, rather than rediscover where things are explained. 

The diagrams help a lot. I found https://plantuml.com/sequence-diagram to be great in this regard: we can check in their source and comment on individual lines, and keep it in sync with the spec. We could have similar ones depicting the parent and child subnets side by side, with messages going back and forth.

Reply:
Noted. I'll address all outstanding comments and do another pass to adapt to this approach, I like it.

Reply:
Full pass with this in mind done. Resolving the comment but feel free to add any concerns you may have.